<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Detection of subclonal SNVs in deep sequencing experiments.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Detection of subclonal SNVs in deep sequencing experiments.
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;deepSNV&rsquo; version 1.10.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="deepSNV-package.html">deepSNV-package</a></td>
<td>Detection of subclonal SNVs in deep sequencing experiments</td></tr>
<tr><td width="25%"><a href="bam2R.html">bam2R</a></td>
<td>Read nucleotide counts from a .bam file</td></tr>
<tr><td width="25%"><a href="shearwater.html">bbb</a></td>
<td>Bayesian beta-binomal test, codename shearwater</td></tr>
<tr><td width="25%"><a href="bf2Vcf.html">bf2Vcf</a></td>
<td>Function to create a 'VCF' object with variant calls from an array of Bayes factors.</td></tr>
<tr><td width="25%"><a href="consensusSequence-methods.html">consensusSequence</a></td>
<td>Calculate the consensus sequence.</td></tr>
<tr><td width="25%"><a href="consensusSequence-methods.html">consensusSequence-method</a></td>
<td>Calculate the consensus sequence.</td></tr>
<tr><td width="25%"><a href="control-methods.html">control</a></td>
<td>Get control counts</td></tr>
<tr><td width="25%"><a href="control-methods.html">control-method</a></td>
<td>Get control counts</td></tr>
<tr><td width="25%"><a href="coordinates-methods.html">coordinates</a></td>
<td>Get coordinates</td></tr>
<tr><td width="25%"><a href="coordinates-methods.html">coordinates-method</a></td>
<td>Get coordinates</td></tr>
<tr><td width="25%"><a href="deepSNV-methods.html">deepSNV</a></td>
<td>Test two matched deep sequencing experiments for low-frequency SNVs.</td></tr>
<tr><td width="25%"><a href="deepSNV-class.html">deepSNV-class</a></td>
<td>deepSNV class.</td></tr>
<tr><td width="25%"><a href="deepSNV-methods.html">deepSNV-method</a></td>
<td>Test two matched deep sequencing experiments for low-frequency SNVs.</td></tr>
<tr><td width="25%"><a href="estimateDirichlet-methods.html">estimateDirichlet</a></td>
<td>Learn a base-specific Dirichlet prior.</td></tr>
<tr><td width="25%"><a href="estimateDirichlet-methods.html">estimateDirichlet-method</a></td>
<td>Learn a base-specific Dirichlet prior.</td></tr>
<tr><td width="25%"><a href="estimateDispersion-methods.html">estimateDispersion</a></td>
<td>Estimate the Dispersion factor in a beta-binomial model.</td></tr>
<tr><td width="25%"><a href="estimateDispersion-methods.html">estimateDispersion-method</a></td>
<td>Estimate the Dispersion factor in a beta-binomial model.</td></tr>
<tr><td width="25%"><a href="estimateRho.html">estimateRho</a></td>
<td>Helper function for estimating the dispersion factor rho</td></tr>
<tr><td width="25%"><a href="Extract-methods.html">Extract</a></td>
<td>Subsetting for deepSNV objects.</td></tr>
<tr><td width="25%"><a href="trueSNVs.html">HIVmix</a></td>
<td>Example .bam data and true SNVs.</td></tr>
<tr><td width="25%"><a href="loadAllData.html">loadAllData</a></td>
<td>Function to load all data from a list of bam files</td></tr>
<tr><td width="25%"><a href="makePrior.html">makePrior</a></td>
<td>Compute a prior from a COSMIC VCF object</td></tr>
<tr><td width="25%"><a href="manhattanPlot.html">manhattanPlot</a></td>
<td>Manhattan plot.</td></tr>
<tr><td width="25%"><a href="mcChunk.html">mcChunk</a></td>
<td>Little helper function to split the count objects into a smaller digestible chunks and run function FUN on each subset</td></tr>
<tr><td width="25%"><a href="normalize-methods.html">normalize</a></td>
<td>Normalize nucleotide counts.</td></tr>
<tr><td width="25%"><a href="normalize-methods.html">normalize-method</a></td>
<td>Normalize nucleotide counts.</td></tr>
<tr><td width="25%"><a href="p.combine.html">p.combine</a></td>
<td>Combine two p-values</td></tr>
<tr><td width="25%"><a href="p.val-methods.html">p.val</a></td>
<td>Get p-values</td></tr>
<tr><td width="25%"><a href="p.val-methods.html">p.val-method</a></td>
<td>Get p-values</td></tr>
<tr><td width="25%"><a href="phiX.html">phiX</a></td>
<td>Example phiX data</td></tr>
<tr><td width="25%"><a href="pi.html">pi</a></td>
<td>Example prior</td></tr>
<tr><td width="25%"><a href="plot.deepSNV.html">plot.deepSNV</a></td>
<td>Scatter plot of relative nucleotide frequencies.</td></tr>
<tr><td width="25%"><a href="RCC.html">RCC</a></td>
<td>Example RCC data</td></tr>
<tr><td width="25%"><a href="repeatMask-methods.html">repeatMask</a></td>
<td>Mask homopolymeric repeats.</td></tr>
<tr><td width="25%"><a href="repeatMask-methods.html">repeatMask-method</a></td>
<td>Mask homopolymeric repeats.</td></tr>
<tr><td width="25%"><a href="RF.html">RF</a></td>
<td>Relative frequencies.</td></tr>
<tr><td width="25%"><a href="shearwater.html">shearwater</a></td>
<td>Bayesian beta-binomal test, codename shearwater</td></tr>
<tr><td width="25%"><a href="show.html">show</a></td>
<td>Show method for deepSNV objects</td></tr>
<tr><td width="25%"><a href="show.html">show-method</a></td>
<td>Show method for deepSNV objects</td></tr>
<tr><td width="25%"><a href="summary-methods.html">summary</a></td>
<td>Summary of a deepSNV object</td></tr>
<tr><td width="25%"><a href="summary-methods.html">summary-method</a></td>
<td>Summary of a deepSNV object</td></tr>
<tr><td width="25%"><a href="test-methods.html">test</a></td>
<td>Get test counts</td></tr>
<tr><td width="25%"><a href="test-methods.html">test-method</a></td>
<td>Get test counts</td></tr>
<tr><td width="25%"><a href="trueSNVs.html">trueSNVs</a></td>
<td>Example .bam data and true SNVs.</td></tr>
<tr><td width="25%"><a href="Extract-methods.html">[-method</a></td>
<td>Subsetting for deepSNV objects.</td></tr>
</table>
</body></html>
