<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: deepSNV class.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for deepSNV-class {deepSNV}"><tr><td>deepSNV-class {deepSNV}</td><td align="right">R Documentation</td></tr></table>

<h2>deepSNV class.</h2>

<h3>Description</h3>

<p>This class stores the contents of the deepSNV test. It is
typically initialized with <code><a href="deepSNV-methods.html">deepSNV</a></code>. This
class has the following slots: </p>

<dl>
<dt>p.val</dt><dd><p>The P-values of the test.</p>
</dd> <dt>test</dt><dd><p>A
matrix with the nucleotide counts in the test experiment.
The column names of the nucleotide counts are A, T, C, G,
- for the positivie strand and a, t, c, g, _ for the
reverse.</p>
</dd> <dt>control</dt><dd><p>A matrix with the nucleotide
counts in the control experiment. The column names must
be the same as for the test.</p>
</dd> <dt>coordinates</dt><dd><p>A
<code><a href="../../base/html/data.frame.html">data.frame</a></code> with the genomic coordinates chr
and pos, and other columns, if desired.</p>
</dd>
<dt>dirichlet.prior</dt><dd><p>A matrix with the
nucleotide-specific Dirichlet prior</p>
</dd>
<dt>pseudo.count</dt><dd><p>The pseudo count if used)</p>
</dd>
<dt>alternative</dt><dd><p>A string with the alternative used in
the test.</p>
</dd> <dt>nucleotides</dt><dd><p>A character vector with the
nucleotides tested.</p>
</dd> <dt>regions</dt><dd><p>A
<code><a href="../../base/html/data.frame.html">data.frame</a></code> with columns chr, start, and
stop.</p>
</dd> <dt>files</dt><dd><p>A list with two entries test and
control storing the filenames (if the object was
initialized from two bam-files).</p>
</dd>
<dt>combine.method</dt><dd><p>The method for combining p-values
as a character string.</p>
</dd> <dt>model</dt><dd><p>The statistical
model, either bin for binomial, or betabin for
beta-binomial</p>
</dd> <dt>over.dispersion</dt><dd><p>If the model is
beta-binomial, the first parameter for the beta-binomial
model, which is shared across sites.</p>
</dd> <dt>call</dt><dd><p>The
last function call to deepSNV.</p>
</dd> <dt>log.lik</dt><dd><p>The log
likelihood of the data under the null hypothesis.
(Excluding zeros on the opposite site under a one-sided
test.)</p>
</dd> </dl>



<h3>Author(s)</h3>

<p>Moritz Gerstung
</p>


<h3>See Also</h3>

<p><code><a href="deepSNV-methods.html">deepSNV</a></code>
</p>


<h3>Examples</h3>

<pre>
## Short example with 2 SNVs at frequency ~10%
regions &lt;- data.frame(chr="B.FR.83.HXB2_LAI_IIIB_BRU_K034", start = 3120, stop=3140)
ex &lt;- deepSNV(test = system.file("extdata", "test.bam", package="deepSNV"), control = system.file("extdata", "control.bam", package="deepSNV"), regions=regions, q=10)
show(ex)   # show method
plot(ex)   # scatter plot
summary(ex)   # summary with significant SNVs
ex[1:3,]   # subsetting the first three genomic positions
tail(test(ex, total=TRUE))   # retrieve the test counts on both strands
tail(control(ex, total=TRUE))

## Not run: Full example with ~ 100 SNVs. Requires an internet connection, but try yourself.
# regions &lt;- data.frame(chr="B.FR.83.HXB2_LAI_IIIB_BRU_K034", start = 2074, stop=3585)
# HIVmix &lt;- deepSNV(test = "http://www.bsse.ethz.ch/cbg/software/deepSNV/data/test.bam", control = "http://www.bsse.ethz.ch/cbg/software/deepSNV/data/control.bam", regions=regions, q=10)
data(HIVmix) # attach data instead..
show(HIVmix)
plot(HIVmix)
head(summary(HIVmix))

</pre>

<hr><div align="center">[Package <em>deepSNV</em> version 1.10.0 <a href="00Index.html">Index</a>]</div>
</body></html>
