<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Test two matched deep sequencing experiments for...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for deepSNV {deepSNV}"><tr><td>deepSNV {deepSNV}</td><td align="right">R Documentation</td></tr></table>

<h2>Test two matched deep sequencing experiments for low-frequency SNVs.</h2>

<h3>Description</h3>

<p>This generic function can handle different types of
inputs for the test and control experiments. It either
reads from two .bam files, uses two matrices of
nucleotide counts, or re-evaluates the test results from
a <code><a href="deepSNV-class.html">deepSNV-class</a></code> object. The actual test is
a likelihood ratio test of a (beta-)binomial model for
the individual nucleotide counts on each position under
the hypothesis that both experiments share the same
parameter, and the alternative that the parameters
differ. Because the difference in degrees of freedom is
1, the test statistic <i>D = -2 \log
  \max{L_0}/\max{L_1}</i> is asymptotically distributed as
<i>&chi;_1^2</i>. The statistic may be tuned by a
nucleotide specific Dirichlet prior that is learned
across all genomic sites, see
<code><a href="estimateDirichlet-methods.html">estimateDirichlet</a></code>. If the model is
beta-binomial, a global dispersion parameter is used for
all sites. It can be learned with
<code><a href="estimateDispersion-methods.html">estimateDispersion</a></code>.
</p>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>test</code></td>
<td>
<p>The test experiment. Either a .bam file, or a
matrix with nucleotide counts, or a
<code><a href="deepSNV-class.html">deepSNV-class</a></code> object.</p>
</td></tr>
<tr valign="top"><td><code>control</code></td>
<td>
<p>The control experiment. Must be of the
same type as test, or missing if test is a
<code><a href="deepSNV-class.html">deepSNV-class</a></code> object.</p>
</td></tr>
<tr valign="top"><td><code>alternative</code></td>
<td>
<p>The alternative to be tested. One of
greater, less, or two.sided.</p>
</td></tr>
<tr valign="top"><td><code>model</code></td>
<td>
<p>Which model to use. Either &quot;bin&quot;, or
&quot;betabin&quot;. Default &quot;bin&quot;.</p>
</td></tr>
<tr valign="top"><td><code>dirichlet.prior</code></td>
<td>
<p>A base-sepecific Dirichlet prior
specified as a matrix. Default NULL.</p>
</td></tr>
<tr valign="top"><td><code>pseudo.count</code></td>
<td>
<p>If dirichlet.prior=NULL, a
pseudocount can be used to define a flat prior.</p>
</td></tr>
<tr valign="top"><td><code>over.dispersion</code></td>
<td>
<p>A numeric factor for the
over.dispersion, if the model is beta-binomial. Default
100.</p>
</td></tr>
<tr valign="top"><td><code>combine.method</code></td>
<td>
<p>The method to combine p-values. One
of &quot;fisher&quot; (default), &quot;max&quot;, or &quot;average&quot;. See
<code><a href="p.combine.html">p.combine</a></code> for details.</p>
</td></tr>
<tr valign="top"><td><code>regions</code></td>
<td>
<p>The regions to be parsed if test and
control are .bam files. Either a <code><a href="../../base/html/data.frame.html">data.frame</a></code>
with columns &quot;chr&quot; (chromosome), &quot;start&quot;, &quot;stop&quot;, or a
<code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> object. If multiple regions are
specified, the appropriate slots of the returned object
are concatenated by row.</p>
</td></tr>
<tr valign="top"><td><code>q</code></td>
<td>
<p>The quality arguement passed to
<code><a href="bam2R.html">bam2R</a></code> if the experiments are .bam files.</p>
</td></tr>
<tr valign="top"><td><code>s</code></td>
<td>
<p>The strand argument passed to
<code><a href="bam2R.html">bam2R</a></code> if the experiments are .bam files.</p>
</td></tr>
<tr valign="top"><td><code>head.clip</code></td>
<td>
<p>The head.clip argument passed to
<code><a href="bam2R.html">bam2R</a></code> if the experiments are .bam files.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A <code><a href="deepSNV-class.html">deepSNV</a></code> object
</p>


<h3>Author(s)</h3>

<p>Moritz Gerstung
</p>


<h3>Examples</h3>

<pre>
## Short example with 2 SNVs at frequency ~10%
regions &lt;- data.frame(chr="B.FR.83.HXB2_LAI_IIIB_BRU_K034", start = 3120, stop=3140)
ex &lt;- deepSNV(test = system.file("extdata", "test.bam", package="deepSNV"), control = system.file("extdata", "control.bam", package="deepSNV"), regions=regions, q=10)
show(ex)   # show method
plot(ex)   # scatter plot
summary(ex)   # summary with significant SNVs
ex[1:3,]   # subsetting the first three genomic positions
tail(test(ex, total=TRUE))   # retrieve the test counts on both strands
tail(control(ex, total=TRUE))

## Not run: Full example with ~ 100 SNVs. Requires an internet connection, but try yourself.
# regions &lt;- data.frame(chr="B.FR.83.HXB2_LAI_IIIB_BRU_K034", start = 2074, stop=3585)
# HIVmix &lt;- deepSNV(test = "http://www.bsse.ethz.ch/cbg/software/deepSNV/data/test.bam", control = "http://www.bsse.ethz.ch/cbg/software/deepSNV/data/control.bam", regions=regions, q=10)
data(HIVmix) # attach data instead..
show(HIVmix)
plot(HIVmix)
head(summary(HIVmix))

</pre>

<hr><div align="center">[Package <em>deepSNV</em> version 1.10.0 <a href="00Index.html">Index</a>]</div>
</body></html>
