<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Detection of subclonal SNVs in deep sequencing experiments</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for deepSNV-package {deepSNV}"><tr><td>deepSNV-package {deepSNV}</td><td align="right">R Documentation</td></tr></table>

<h2>Detection of subclonal SNVs in deep sequencing experiments</h2>

<h3>Description</h3>

<p>Detection of subclonal SNVs in deep sequencing
experiments
</p>


<h3>Details</h3>

<p>This packages provides algorithms for detecting subclonal
single nucleotide variants (SNVs) and their frequencies
from ultra-deep sequencing data. It retrieves the
nucleotide counts at each position and each strand from
two .bam files and tests for differences between the two
experiments with a likelihood ratio test using either a
binomial or and overdispersed beta-binomial model. The
statistic can be tuned across genomic sites by a shared
Dirichlet prior and there package provides procedures for
normalizing sequencing data from different runs.
</p>


<h3>Author(s)</h3>

<p>Moritz Gerstung, Wellcome Trust Sanger Institute,
<a href="mailto:moritz.gerstung@sanger.ac.uk">moritz.gerstung@sanger.ac.uk</a>
</p>


<h3>References</h3>

<p>Gerstung M, Beisel C, Rechsteiner M, Wild P, Schraml P,
Moch H, and Beerenwinkel N. Reliable detection of
subclonal single-nucleotide variants in tumour cell
populations. Nat Commun 3:811 (2012).
<a href="http://dx.doi.org/10.1038/ncomms1814">DOI:10.1038/ncomms1814</a>.
</p>


<h3>See Also</h3>

<p><code><a href="deepSNV-methods.html">deepSNV</a></code>
</p>


<h3>Examples</h3>

<pre>
## Short example with 2 SNVs at frequency ~10%
regions &lt;- data.frame(chr="B.FR.83.HXB2_LAI_IIIB_BRU_K034", start = 3120, stop=3140)
ex &lt;- deepSNV(test = system.file("extdata", "test.bam", package="deepSNV"), control = system.file("extdata", "control.bam", package="deepSNV"), regions=regions, q=10)
show(ex)   # show method
plot(ex)   # scatter plot
summary(ex)   # summary with significant SNVs
ex[1:3,]   # subsetting the first three genomic positions
tail(test(ex, total=TRUE))   # retrieve the test counts on both strands
tail(control(ex, total=TRUE))

## Not run: Full example with ~ 100 SNVs. Requires an internet connection, but try yourself.
# regions &lt;- data.frame(chr="B.FR.83.HXB2_LAI_IIIB_BRU_K034", start = 2074, stop=3585)
# HIVmix &lt;- deepSNV(test = "http://www.bsse.ethz.ch/cbg/software/deepSNV/data/test.bam", control = "http://www.bsse.ethz.ch/cbg/software/deepSNV/data/control.bam", regions=regions, q=10)
data(HIVmix) # attach data instead..
show(HIVmix)
plot(HIVmix)
head(summary(HIVmix))

</pre>

<hr><div align="center">[Package <em>deepSNV</em> version 1.10.0 <a href="00Index.html">Index</a>]</div>
</body></html>
